text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+imported from NPM, and we're trying to wrap it with a `TimerComponent` that
+uses it to keep track of how much time has passed:
+
+```js
+// simple-timer/index.js
+export default class Timer {
+  seconds = 0;
+  minutes = 0;
+  hours = 0;
+
+  listeners = [];
+
+  constructor() {
+    setInterval(() => {
+      this.seconds++;
+      this.minutes = Math.floor(this.seconds / 60);
+      this.hours = Math.floor(this.minutes / 60);
+      this.notifyTick();
+    }, 1000);
+  }
+
+  notifyTick() {
+    for (let listener of this.listeners) {
+      listener(this.seconds);
+    }
+  }
+
+  onTick(listener) {
+    this.listeners.push(listener);
+  }
+}
+```
+
+```js
+import Timer from 'simple-timer';
+import Component, { tracked } from '@glimmer/tracking';
+
+export default class TimerComponent extends Component {
+  @tracked timer = new Timer();
+
+  @tracked
+  get currentSeconds() {
+    return this.timer.seconds;
+  }
+
+  @tracked
+  get currentMinutes() {
+    return this.timer.minutes;
+  }
+}
+```
+
+Even though we've marked the `timer` property as tracked, the `timer.seconds`
+property is untracked, and _it_ is the field that is updated. We can solve this
+problem by using the timer library's `onTick` event handler to re-set the field,
+invalidating it:
+
+```js
+export default class TimerComponent extends Component {
+  @tracked timer = new Timer();
+
+  constructor() {
+    this.timer.onTick(() => {
+      // invalidate the timer field.
+      this.timer = this.timer;
+    });
+  }
+
+  @tracked
+  get currentSeconds() {
+    return this.timer.seconds;
+  }
+
+  @tracked
+  get currentMinutes() {
+    return this.timer.minutes;
+  }
+}
+```
+
+### Interop with the Classic Programming Model
+
+Tracked properties represent a paradigm shift. They are a completely new system,
+fully independent of the classic programming model and based on modern
+Javascript features and design, and they will be the _default_ change tracking
+system in Ember Octane.
+
+However, existing apps, libraries, and addons will be using the classic
+programming model for some time, and experience tells us that these sort of
+transitions to new features take a while to settle in the community. To ease
+this process and enable gradual adoption, tracked properties will be able to
+interoperate with the most commonly used features of the classic model:
+
+- Classic classes
+- Computed properties
+- `get`/`set` and property notifications
+
+Tracked properties will _not_ interoperate with observers, which are strictly
+within the old paradigm.
+
+#### Classic Classes
+
+The `tracked` decorator function will be usable in classic classes, similar to
+`computed`:
+
+```js
+import EmberObject from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+
+const Person = EmberObject.extend({
+  firstName: tracked({ value: 'Tom' }),
+  lastName: tracked({ value: 'Dale' }),
+
+  fullName: tracked({
+    get() {
+      return `${this.firstName} ${this.lastName}`;
+    },
+  }),
```

Comment:
```js
  fullName: tracked(function() {
    return `${this.firstName} ${this.lastName}`;
  }),
```

is this form also supported to make the transition from regular `computed` properties easier?

Reply:
I think this would be reasonable, with the caveat that this is effectively defining a property with _just_ a `get` (no overridability).

Reply:
if we don't cache `tracked` value, we can rewrite it like:

```js
 get fullName() {
    return `${this.firstName} ${this.lastName}`;
 },
```

Reply:
@lifeart This would mean, that if you try to render `{{this.fullName}}` in a template and update `firstName`, Glimmer would have no way of knowing that and would consequently not update `{{this.fullName}}`.

`@tracked` is not about caching, but about change tracking / invalidation.

Reply:
Right. Specifically this becomes a problem when we access values using _plain javascript_, because those accessor have not be instrumented.

Consider this class:

```js
class Person {
  get fullName() {
    return `${this.firstName} ${this.lastName}`;
  }

  get greeting() {
    return `Hello, ${this.fullName}`;
  }
}
```

And this template:

```hbs
{{greeting}}
```

Glimmer knows that the path `greeting` is being read from. It could in theory add some method of tracking this change, like a native setter (this would be _very_ expensive, but possible). However, when we _get_ `greeting`, Glimmer has no way to know what's happening internally. When `greeting` accesses `this.fullName`, that's a standard JS lookup, and it's fully opaque to the outside world.

We need to add something which can _intercept_ this access, and that's what `@tracked` is doing. Every tracked property intercepts its access, and pushes that onto a stack. At the end of getting `greeting`, we know that `fullName` and `firstName` and `lastName` were all accessed.

`@tracked` also intercepts updates, so later on when we set `firstName` it bumps `firstName`s revision counter. If we didn't decorate the property, we would have no way to intercept the update either (without a method like `set`).

Reply:
looks creepy, why imho native getters must work out of the box like computed volatile

Reply:
@lifeart they _do_ work OOTB, just like computed volatile, they just don't have any method of invalidating when they changed. I understand wanting native getters and setters to Just Work™️, I spent quite a while deep diving into native Proxy to see if there would be any way to do it in the future too, but so far I do not think it is possible. Specifically, any Proxies will break when private fields are adopted. This was a _huge_ issue for Vue and Aurelia maintainers, they're still in talks with TC39 over it (you can see their conversation [here](https://github.com/tc39/proposal-class-fields/issues/106)).

If TC39 ever decides that full object wrapping/getter interception is a priority and can be added to the language, I think it opens up quite a few possibilities, but at the moment we are fairly limited here.

Reply:
@pzuraq is it possible to invalidate it on each rerender automatically? (or make tags dirty after each render)

Reply:
@lifeart why would you want that?

Reply:
If we did that, then we would be rerendering _everything_ every render, right? We need to know what mutated and changed, and the only way we can tell that is by marking it as change _when_ it changes.

In React, they use `setState` and `shouldComponentUpdate` (I believe, not a React expert). These let React know "hey, something changed here, you should update this component tree". In the same way, we need to let Glimmer know something changed by marking a tag as dirtied when it changes.

Reply:
re: react - shouldComponentUpdate can _also_ be used to determine when props (args to a component) should cause an update / re-render.

Reply:
@NullVoxPopuli I need behaviour like `helpers` (always recalculating on rerender) and it will be nice to have it by default for properties with defined getters and without `@tracked` decorator

Reply:
> For example, if I need to return random value on each rerender,
> get random() { return Math.random(); } going to be tricky with glimmer

do you mean you just want a property that isn't tracked, so that every render, the property is evaluated for the render? (not evaluated as a consequence of the render)

Reply:
@NullVoxPopuli yes, just like `helper` in handlebars, we need to call it on each rerender.

Reply:
so... that has nothing to do with how tracked works? I am confused

Reply:
@NullVoxPopuli now if you have defined property getter without `@tracked`, first calclulation result will be always cached in view. (no public way to invalidate it)

Reply:
so the issue is;
 - all properties are cached, regardless of trackedness
 - even if _something_ causes a re-render, the non-tracked property is not re-evaluated

Reply:
I’m struggling to think of a valid use case for wanting a property to update _only in response to a render cycle_, since this is quite unpredictable. I can’t think of any UI that wants to update over time that would be better modeled via random opportunistic updates during rerenders than something more deterministic like scheduling updates with a timer.

Reply:
How about “currentTimestamp” property, returning Date.now()?

Reply:
Let's assume we want to have component with  "tick" programming model.

somehow we update "iteration" property, and other values recalculated like computed volatile.

```js
{

	@tracked iteration = 0;

	didUpdate() {
		setTimeout(()=>{
			this.iteration++;
		}, 500);
	}

	get currentTimestamp() {
		return Date.now();
	}

	get availableSections() {
		return Math.random > 0.5 ? true : false;
	}

}
```

of course, we can rewrite it like:

```js
{

	@tracked currentTimestamp = Date.now();
	@tracked hasAvailableSections = Math.random > 0.5 ? true : false;

	didUpdate() {
		setTimeout(()=>{
			this.currentTimestamp = Date.now();
			this.availableSections = Math.random > 0.5 ? true : false;
		}, 500);
	}
}
```

but it's looks more overheaded